pkg/controller/logstash/service.go

```diff
@@ -0,0 +1,92 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package logstash
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	logstashv1alpha1 "github.com/elastic/cloud-on-k8s/v2/pkg/apis/logstash/v1alpha1"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/common"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/common/defaults"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/logstash/network"
+)
+
+// reconcileServices reconcile Services defined in the spec.
+// When Services are empty, a default Service for port 9600 is created.
+// If api.http.port is customized, user is expected to config Services.
+// When Services exist, the port 9600 does not attach to any of Service.
+func reconcileServices(params Params) ([]corev1.Service, error) {
+	if len(params.Logstash.Spec.Services) == 0 {
```

Comment:
Rather than checking if the number of services is 0, we should probably check whether a service has been spec'd with the same name as the `default` service - I think it may make sense to rename this to `api`, or `internal`, rather than `default`, WDYT?

Reply:
we can make a default headless service regardless of user input similar to ES
```
NAME                                    TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)    AGE
elasticsearch-sample-es-default         ClusterIP   None            <none>        9200/TCP   17h
elasticsearch-sample-es-http            ClusterIP   10.98.229.240   <none>        9200/TCP   17h
elasticsearch-sample-es-internal-http   ClusterIP   10.96.255.68    <none>        9200/TCP   17h
elasticsearch-sample-es-transport       ClusterIP   None            <none>        9300/TCP   17h
```
For user defined service, name it with suffix`api`.  If user doesn't provide input, create a `ClusterIP` service with suffix `http` for metrics.

Reply:
I was wondering if we could call the service for metrics `api`,`internal` or `http`, regardless of whether they have set values in the `spec`. And if they define a service with the same name, it overrides the default settings, allowing the user to change the cluster type from the default if they need to - say if they have an external monitoring tool that reads from the logstash API.

Reply:
yes. Using the name as an identifier to override default settings (eg. port) sounds good. 

My concern is this [comment](https://github.com/elastic/cloud-on-k8s/pull/6404#discussion_r1111119489) 
> K8s docs ask for a headless service for each stateful set

Regardless of user's setting in `spec`, I think we can always create headless service for Logstash

Reply:
Yes, I agree - I do think we need to always create a headless service